### PR TITLE
PeriodicVariables2 Noelle PR

### PR DESCRIPTION
**Pull request by dmdlott**
PeriodicVariableSCC now more carefully chooses its accumulator.

Also adds detection for 2-phis-case Periodic Variable SCC, provided only one of the two phis has SCC-external users.

This PR is a sister to the GINO PR fixing PeriodicVariables2.

### CODE REVIEW

File: src/core/loop_scc_attributes/src/PeriodicVariableSCC.cpp [hunk 1]
```diff
@@ -40,6 +41,18 @@ PeriodicVariableSCC::PeriodicVariableSCC(
     period{ per },
     step{ st } {
 
+  // DDLOTT: the algorithm for determining the accumulator in
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/src/PeriodicVariableSCC.cpp:44]: This comment is confusing. The class SingleAccumulatorRecomputableSCC is meant to handle a single-accumulator SCC as the name suggests. So it's strange to have a comment in this class saying the algorithm included within is not sufficient for 2-phis (hence multiple accumulators).

File: src/core/loop_scc_attributes/src/PeriodicVariableSCC.cpp [hunk 2]
```diff
@@ -40,6 +41,18 @@ PeriodicVariableSCC::PeriodicVariableSCC(
...
+  // SingleAccumulatorRecomputableSCC is not sufficient for 2-phis case PVSCC.
+  // The representation of SARSCC is kind of fine, since only 1 phi is allowed
+  // to have non-phi users, but the algorithm can't deal with picking the phi
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/src/PeriodicVariableSCC.cpp:47]: I didn't get this comment. All phis have users.

File: src/core/loop_scc_attributes/src/PeriodicVariableSCC.cpp [hunk 3]
```diff
@@ -40,6 +41,18 @@ PeriodicVariableSCC::PeriodicVariableSCC(
...
+  // that has users. errs() << "PV constructor\n";
+  if (acc != nullptr) {
+    // DD: responsibility for this being a PHINode lies with the SCCDAGAttrs
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/src/PeriodicVariableSCC.cpp:50]: Please remove "DD".

I didn't get this comment as well.

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 4]
```diff
@@ -118,8 +120,13 @@ SCCDAGAttrs::SCCDAGAttrs(bool enableFloatAsReal,
 
     } else if (std::get<0>(isPeriodic)) {
       auto loopCarriedDependences = this->sccToLoopCarriedDependencies.at(scc);
-      Value *initialValue, *period, *step;
-      tie(std::ignore, initialValue, period, step) = isPeriodic;
+      // DDLOTT Sep/3/2024: SingleAccumulatorRecomputableSCC's accumulator
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:123]: Please remove "DDLOT" and the date.

Comment (scampanoni) [reply src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:123]: I didn't get this comment

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 5]
```diff
@@ -541,10 +549,10 @@ std::set<InductionVariable *> SCCDAGAttrs::
   return containedIVs;
 }
 
-std::tuple<bool, Value *, Value *, Value *> SCCDAGAttrs::checkIfPeriodic(
-    SCC *scc,
-    LoopTree *loopNode) {
-  auto notPeriodic = std::make_tuple(false, nullptr, nullptr, nullptr);
+std::tuple<bool, Value *, Value *, Value *, Value *> SCCDAGAttrs::
+    checkIfPeriodic(SCC *scc, LoopTree *loopNode) {
+  // DD: whether or not a variable is periodic is not decidable.
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:554]: Please remove "DD"

Comment (scampanoni) [reply src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:554]: What is the purpose of this comment?

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 6]
```diff
@@ -569,17 +577,91 @@ std::tuple<bool, Value *, Value *, Value *> SCCDAGAttrs::checkIfPeriodic(
     Value *initialValue;
     Value *period;
     Value *step;
+    Value *accumulator; // should be nullptr at return unless the Periodic
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:580]: Please initialize it to nullptr

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 7]
```diff
@@ -569,17 +577,91 @@ std::tuple<bool, Value *, Value *, Value *> SCCDAGAttrs::checkIfPeriodic(
...
 
     if (!isa<PHINode>(to))
       return notPeriodic;
+
     auto toPHI = cast<PHINode>(to);
 
     if (toPHI->getNumIncomingValues() != 2)
       return notPeriodic;
 
+    // DD: we want to capture a case where we have phi1=(ph, -x)(latch, phi2)
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:594]: Remove "DD"

Comment (scampanoni) [reply src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:594]: Please expand the code in the comment aiming for higher precision.

**Review by scampanoni — CHANGES_REQUESTED** [review]
The code looks good to me (great job!).

The comments need to improve before we merge this PR.

**Comment by dmdlott** [comment]
@scampanoni bumping this

File: src/core/loop_scc_attributes/src/PeriodicVariableSCC.cpp [hunk 8]
```diff
@@ -41,17 +41,22 @@ PeriodicVariableSCC::PeriodicVariableSCC(
...
-    // isPeriodic analysis. errs()  << *acc << "\n" << this->accumulator <<
-    // "\n";
-    this->accumulator = cast<PHINode>(acc);
-  }
+      /*
+       * We allow PeriodicVariableSCC to include cases that feature a single SCC containing
+       * two phis when one of the phis is used only by the other phi.
+       * We view this as a single accumulator case where the accumulator is said to be the phi 
+       * that has SCC-external users, and the existing code that uses PeriodicVariableSCC (which
+       * is in GINO) is fully compatible with that representation.
```
Comment (scampanoni) [inline src/core/loop_scc_attributes/src/PeriodicVariableSCC.cpp:49]: We should not reference code that builds upon NOELLE.

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 9]
```diff
@@ -120,11 +120,14 @@ SCCDAGAttrs::SCCDAGAttrs(bool enableFloatAsReal,
 
     } else if (std::get<0>(isPeriodic)) {
       auto loopCarriedDependences = this->sccToLoopCarriedDependencies.at(scc);
+<<<<<<< HEAD
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:123]: Please delete this

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 10]
```diff
@@ -120,11 +120,14 @@ SCCDAGAttrs::SCCDAGAttrs(bool enableFloatAsReal,
...
+=======
       // DDLOTT Sep/3/2024: SingleAccumulatorRecomputableSCC's accumulator
       // picking algorithm doesn't suffice for 2-phis SCC. Rather than change
       // that algorithm, it's probably better to put the code compensating for
       // that deficiency so that it's attached to PeriodicVariableSCC -- which
       // we might split off of SARSCC in the future.
+>>>>>>> 483b2ea3fe44c205a5f722d79483f07f562938a9
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:130]: Please delete this

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 11]
```diff
@@ -549,9 +552,15 @@ std::set<InductionVariable *> SCCDAGAttrs::
   return containedIVs;
 }
 
+<<<<<<< HEAD
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:555]: You probably merged without checking these files

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 12]
```diff
@@ -577,8 +586,12 @@ std::tuple<bool, Value *, Value *, Value *, Value *> SCCDAGAttrs::
     Value *initialValue;
     Value *period;
     Value *step;
+<<<<<<< HEAD
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:589]: Problem with merging

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 13]
```diff
@@ -549,9 +552,15 @@ std::set<InductionVariable *> SCCDAGAttrs::
...
 
+<<<<<<< HEAD
+std::tuple<bool, Value *, Value *, Value *, Value *> SCCDAGAttrs::checkIfPeriodic(
+    SCC *scc,
+    LoopTree *loopNode) {
+=======
 std::tuple<bool, Value *, Value *, Value *, Value *> SCCDAGAttrs::
     checkIfPeriodic(SCC *scc, LoopTree *loopNode) {
   // DD: whether or not a variable is periodic is not decidable.
+>>>>>>> 483b2ea3fe44c205a5f722d79483f07f562938a9
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:563]: Problem with merging

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 14]
```diff
@@ -577,8 +586,12 @@ std::tuple<bool, Value *, Value *, Value *, Value *> SCCDAGAttrs::
     Value *initialValue;
     Value *period;
     Value *step;
+<<<<<<< HEAD
+    Value *accumulator = nullptr; //should be nullptr at return unless the Periodic Variable contains 2+ PHINodes. Identifies accumulator
+=======
     Value *accumulator; // should be nullptr at return unless the Periodic
                         // Variable contains 2+ PHINodes. Identifies accumulator
+>>>>>>> 483b2ea3fe44c205a5f722d79483f07f562938a9
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:594]: Problem with merging

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 15]
```diff
@@ -591,13 +604,26 @@ std::tuple<bool, Value *, Value *, Value *, Value *> SCCDAGAttrs::
     if (toPHI->getNumIncomingValues() != 2)
       return notPeriodic;
 
+<<<<<<< HEAD
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:607]: Problem with merging

File: src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp [hunk 16]
```diff
@@ -591,13 +604,26 @@ std::tuple<bool, Value *, Value *, Value *, Value *> SCCDAGAttrs::
...
+    if(isa<PHINode>(from)) {
+=======
     // DD: we want to capture a case where we have phi1=(ph, -x)(latch, phi2)
     // phi2=(ph, x)(latch, phi1) This is a different way of expressing a
     // subtract-from-zero flipflop. In such a case, only one of the phis should
     // have scc-external users. The other phi should only be holding the "out of
     // phase" value. If instead the other phi has scc-external users, it implies
     // that the scc is composed of two interdependent variables.
     if (isa<PHINode>(from)) {
+>>>>>>> 483b2ea3fe44c205a5f722d79483f07f562938a9
```
Comment (scampanoni) [inline src/core/loop_sccdag_attributes/src/SCCDAGAttrs.cpp:626]: Problem with merging

**Review by scampanoni — CHANGES_REQUESTED** [review]

**Pull request merged**